File: Tree/538. Convert BST to Greater Tree.cpp
/*
538. Convert BST to Greater Tree

Given a Binary Search Tree (BST), 
convert it to a Greater Tree such that every key of the original BST is changed to the original key plus sum of all keys greater than the original key in BST.

Example:

Input: The root of a Binary Search Tree like this:
              5
            /   \
           2     13

Output: The root of a Greater Tree like this:
             18
            /   \
          20     13


*/



class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        int presum = 0;
        Helper(root, presum);
        return root;
    }
    
    void Helper(TreeNode* root, int& presum){
        if(!root) return;
        Helper(root->right, presum);
        root->val += presum;
        presum = root->val;
        Helper(root->left, presum);
    }
};


class Solution {
    int helper(TreeNode* root, int gs) {
        if(!root) return gs;
        root->val += helper(root->right, gs);
        return helper(root->left, root->val);
    }
public:
    TreeNode* convertBST(TreeNode* root) {
        helper(root,0);
        return root;
    }
};

//Morris Traversal
class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        int presum = 0;
        TreeNode* cur = root;
        while(cur){
            if(!cur->right){
                cur->val += presum;
                presum = cur->val;
                cur = cur->left;
            }else{
                TreeNode* pre = cur->right;
                while(pre->left && pre->left != cur)
                    pre = pre->left;
                if(!pre->left){
                    pre->left = cur;
                    cur = cur->right;
                }else{
                    cur->val += presum;
                    presum = cur->val;
                    cur = cur->left;
                    pre->left = NULL;
                }
            }
            
        }
        return root;
    }
};





//Solution 1: Pointer of Pointer 

class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        TreeNode* pre = nullptr;
        helper(&root, pre);
        return root;
    }
    
    void helper(TreeNode** root, TreeNode*& pre){
        if(!*root) return;
        helper(&((*root)->right), pre);
        if(pre)
            (*root)->val += pre->val;
        pre = *root;
        helper(&((*root)->left), pre);
    }
};

//Solution 2: TreeNode 

class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        TreeNode* pre = nullptr;
        helper(root, pre);
        return root;
    }
    
    void helper(TreeNode* root, TreeNode*& pre){
        if(!root) return ;
        helper(root->right, pre);
        if(pre)
            root->val += pre->val;
        pre = root;
        helper(root->left, pre);
        return;
    }
};


//Solution 3  Morris 
class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        TreeNode* cur = root;
        TreeNode* prev = nullptr;
        while(cur){
            if(!cur->right){
                if(prev) 
                    cur->val += prev->val;
                prev = cur;
                cur = cur->left;
            }else{
                TreeNode* pre = cur->right;
                while(pre->left && pre->left != cur)
                    pre = pre->left;
                if(pre->left){
                    pre->left = NULL;
                    cur->val += prev->val;
                    prev = cur;
                    cur = cur->left;
                }else{
                    pre->left = cur;
                    cur = cur->right;
                }
            }
        }
        return root;
    }
};

File: Tree/538. Convert BST to Greater Tree.py
"""
538. Convert BST to Greater Tree

Input: The root of a Binary Search Tree like this:
              5
            /   \
           2     13

Output: The root of a Greater Tree like this:
             18
            /   \
          20     13

"""


""""
        5 
       /   \ 
      2    13   r = 16, node 13, 
     /    /   \ 
    1    12    16 pre = 0
        pre = 29

"""

class Solution:
    def convertBST(self, root):
        self.pre = 0
        def helper(node):
            if not node: return
            helper(node.right)
            node.val = self.pre + node.val
            self.pre = node.val
            helper(node.left)
        helper(root)
        return root


class Solution:
    def convertBST(self, root):
        def helper(node, pre):
            if not node: return pre
            pre =  helper(node.right, pre)
            node.val += pre
            pre = helper(node.left,node.val)
            return pre
        helper(root, 0)
        return root

class Solution:
    def convertBST(self, root):
        def reverse(root):
            return reverse(root.right) + [root] + reverse(root.left) if root else []
        for a, b in zip(reverse(root), reverse(root)[1:]):
            b.val += a.val
        return root


#BFS
class Solution:
    summ = 0
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        stack = [(root, 0)]
        while stack:
            node, visited = stack.pop()
            if node is None:
                continue
            if visited == 0:
                stack.extend([(node.left, 0), (node, 1), (node.right, 0)])
            else:
                self.summ += node.val
                node.val = self.summ
        return root